ee/clickhouse/migrations/test/test_0026_fix_materialized_window_and_session_ids.py

```diff
@@ -0,0 +1,77 @@
+import importlib
+
+from ee.clickhouse.materialized_columns.columns import materialize
+from ee.clickhouse.util import ClickhouseDestroyTablesMixin, ClickhouseTestMixin
+from posthog.client import sync_execute
+from posthog.conftest import create_clickhouse_tables
+from posthog.settings import CLICKHOUSE_DATABASE
+from posthog.test.base import BaseTest
+
+# Import the migration in this way because it starts with a number
+_0026_fix_materialized_window_and_session_ids = importlib.import_module(
+    "ee.clickhouse.migrations.0026_fix_materialized_window_and_session_ids"
+)
+does_column_exist = _0026_fix_materialized_window_and_session_ids.does_column_exist
+materialize_session_and_window_id = _0026_fix_materialized_window_and_session_ids.materialize_session_and_window_id
+
+
+class TestMigration(ClickhouseTestMixin, ClickhouseDestroyTablesMixin, BaseTest):
+    def setUp(self):
+        self.recreate_database()
+        super().setUp()
+
+        # Ideally, we would set this up by running the other migrations leading up
+        # to this migrations, but it's very slow if we do that. So instead, we drop
+        # the columns that this migration creates
+        sync_execute("ALTER TABLE events DROP COLUMN IF EXISTS $session_id")
+        sync_execute("ALTER TABLE events DROP COLUMN IF EXISTS $window_id")
+        sync_execute("ALTER TABLE sharded_events DROP COLUMN IF EXISTS $session_id")
+        sync_execute("ALTER TABLE sharded_events DROP COLUMN IF EXISTS $window_id")
+
+    def tearDown(self):
+        self.recreate_database()
+
+    def recreate_database(self):
+        sync_execute(f"DROP DATABASE {CLICKHOUSE_DATABASE} SYNC")
+        sync_execute(f"CREATE DATABASE {CLICKHOUSE_DATABASE}")
+        create_clickhouse_tables(0)
+
+    def assert_desired_state(self):
+        self.assertTrue(does_column_exist(CLICKHOUSE_DATABASE, "events", "$session_id"))
+        self.assertTrue(does_column_exist(CLICKHOUSE_DATABASE, "sharded_events", "$session_id"))
+        self.assertTrue(does_column_exist(CLICKHOUSE_DATABASE, "events", "$window_id"))
+        self.assertTrue(does_column_exist(CLICKHOUSE_DATABASE, "sharded_events", "$window_id"))
+
+        self.assertFalse(does_column_exist(CLICKHOUSE_DATABASE, "events", "mat_$session_id"))
+        self.assertFalse(does_column_exist(CLICKHOUSE_DATABASE, "sharded_events", "mat_$session_id"))
+        self.assertFalse(does_column_exist(CLICKHOUSE_DATABASE, "events", "mat_$window_id"))
+        self.assertFalse(does_column_exist(CLICKHOUSE_DATABASE, "sharded_events", "mat_$window_id"))
+
+    def test_columns_already_materialized_prior_to_migration(self):
+        materialize("events", "$session_id")
+        materialize("events", "$window_id")
+
+        materialize_session_and_window_id(CLICKHOUSE_DATABASE)
```

Comment:
Non-blocking

I quite like using call_command for that end to end experience

```
from django.core.management import call_command
```

Reply:
Just spent a bit trying to do this.

The issue I was hitting is that we don't want to run all of the migrations (because that takes forever), so in our setup, we need to fake the other migrations. But for some reason, the clickhouse ORM does not like it when we fake migrations in our test env. 

The issue I'm hitting is that we need the migration table to be setup before faking migrations, otherwise, we get a table doesn't exist error - as expected, but if I create the migration table, I get a `REPLICA_IS_ALREADY_EXIST` error. I think it has something to do with how we handle table replication in our test environment, but I couldn't figure it out.

Reply:
Yeah the replication is tricky. We’ll converge on replication everywhere 🤞 and not have to deal with the settings complexity 🚢